SNLS_NLSPowell.h

```diff
@@ -0,0 +1,729 @@
+            for (int i = 0; i < m_nDim; i++) {
+                // work_arr3 = R \Delta x + Q^T F
+                // where F here is our residual vector
+                work_arr3[i] += qtf[i];
+                sum += (work_arr3[i] * work_arr3[i]);
+            }
+            sum = sqrt(sum);
+
+            double predicted_reduction = 0.0;
+
+            // Compute 2nd power
+            if (sum < m_res) {
+                sum = sum / m_res;
+                predicted_reduction = 1.0 - sum * sum;
+            }
+
+            // Compute the ratio of the actual to the predicted reduction
+            const double ratio = (predicted_reduction > 0.0) ? (actual_reduction / predicted_reduction) : 0.0;
+
+            // Update the step bound
+            if (ratio < 0.1) {
+                m_ncsuc = 0;
+                m_ncfail += 1;
+                m_delta *= 0.5;
+            }
+            else {
+                m_ncfail = 0;
+                m_ncsuc += 1;
+                if (ratio > 0.5 || m_ncsuc > 1) {
+                    m_delta = fmax(m_delta, 2.0 * pnorm);
+                }
+                if (fabs(ratio - 1.0) <= 0.1) {
+                    m_delta = 2.0 * pnorm;
+                }
+            }
+
+            // Test for successful iteration
+            if (ratio >= 1e-4) {
+                // Successful iteration
+                // Update norms
+                double sum = 0.0;
+                for (int i = 0; i < m_nDim; i++) {
+                    sum += (diag[i] * m_x[i]) * (diag[i] * m_x[i]);
+                }
+                m_xnorm = sqrt(sum);
+                m_res = res1;
+                m_nIters += 1;
+            }
+            else {
+                // Iteration failed
+                // revert x back to old values
+                for (int i = 0; i < m_nDim; i++) {
+                    m_x[i] -= work_arr1[i];
+                }
+            }
+
+            // Determine the progress of the iteration
+            m_nslow1 = (actual_reduction >= 0.001) ? 0 : (m_nslow1 + 1);
+            if (jeval) {
+                m_nslow2 += 1;
+            }
+            if (actual_reduction >= 0.1) {
+                m_nslow2 = 0;
+            }
+            // Test for convergence
+            if ((m_delta <= (m_x_tolerance * m_xnorm)) || (m_res < m_f_tolerance)) {
+                m_status = converged;
+#ifdef __cuda_host_only__
+                if ( m_os != nullptr ) {
+                    *m_os << "converged" << std::endl ;
+                }
+#endif
+                return m_status;
+            }
+
+            // Tests for termination and stringent tolerances
+            if (m_fevals >= m_maxfev) {
+                m_status = unConvergedMaxIter;
+                return m_status;
+            }
+            if (0.1 * fmax(0.1 * m_delta, pnorm) <= epsmch * m_xnorm) {
+                m_status = deltaFailure;
+                return m_status;
+            }
+            if (m_nslow2 == 5) {
```

Comment:
Back to the copyright question -- this looks like the kind of detail that, if the same as in MINPACK or elsewhere, might trigger a need to be careful about release and licensing details.

Reply:
So, these values are directly defined in the MINPACK manual. So, it would not be a copyright issue as we're just following their API. 

Additionally, I'm going to refactor the PR a bit as I've finally got my hands on the original conference proceeding's paper/chapter that this was algorithm was based on. I actually like the original a bit more as it shares some commonality with our trust-region dog-leg solver. My plans are therefore to do a bit of a hybrid where we just re-use our https://github.com/LLNL/SNLS/blob/develop/SNLS_TrDLDenseG.h#L37-L216 class here to control how delta grows and shrinks. Next, I'll keep the minpack logic for how to exit early if the solver is making no progress as well as their improvement on the original algorithm of doing a rank-1 update of the Jacobian using the Broyden method rather than what appears to be an update of a finite difference method given in the original paper.